scripts/assemble_changelog.py

```diff
@@ -0,0 +1,284 @@
+        for line in input_stream:
+            level, content = self.title_level(line)
+            if level == 2:
+                level_2_seen += 1
+                if level_2_seen <= 1:
+                    self.header.append(line)
+                else:
+                    self.trailer.append(line)
+            elif level == 3 and level_2_seen == 1:
+                current_section = content
+                self.add_sections(current_section)
+            elif level_2_seen == 1 and current_section != None:
+                if line.strip():
+                    self.section_content[current_section].append(line)
+            elif level_2_seen <= 1:
+                self.header.append(line)
+            else:
+                self.trailer.append(line)
+
+    def add_file(self, input_stream):
+        """Add changelog entries from a file.
+
+        Read lines from input_stream, which is typically a file opened
+        for reading. These lines must contain a series of level 3
+        Markdown sections with recognized titles. The corresponding
+        content is injected into the respective sections in the changelog.
+        The section titles must be either one of the hard-coded values
+        in STANDARD_SECTIONS in assemble_changelog.py or already present
+        in ChangeLog.md. Section titles must match byte-for-byte except that
+        leading or trailing whitespace is ignored.
+        """
+        filename = input_stream.name
+        current_section = None
+        for line_number, line in enumerate(input_stream, 1):
+            if not line.strip():
+                continue
+            level, content = self.title_level(line)
+            if level == 3:
+                current_section = content
+                if current_section not in self.section_content:
+                    raise InputFormatError(filename, line_number,
+                                           'Section {} is not recognized',
+                                           str(current_section)[1:])
+            elif level == 0:
+                if current_section is None:
+                    raise InputFormatError(filename, line_number,
+                                           'Missing section title at the beginning of the file')
+                self.section_content[current_section].append(line)
+            else:
+                raise InputFormatError(filename, line_number,
+                                       'Only level 3 headers (###) are permitted')
+
+    def write(self, filename):
+        """Write the changelog to the specified file.
+        """
+        with open(filename, 'wb') as out:
+            for line in self.header:
+                out.write(line)
+            for section in self.section_list:
+                lines = self.section_content[section]
+                while lines and not lines[0].strip():
+                    del lines[0]
+                while lines and not lines[-1].strip():
+                    del lines[-1]
```

Comment:
Suggestion: I think the above 5 lines could be shortened to
```
lines = [l for self.section_content[section] if l.strip()]
```

Also, I was under the impression that we eliminated blank lines when reading, but I don't have anything against a belt-and-braces approach here.

Reply:
The code only strips leading and trailing blank line, not blank lines in the middle. However, since we strip blank lines from the input anyway, that doesn't matter. I'll simplify the code.

Reply:
Good point, I'm not sure how I missed that. Glad the simplification applied nonetheless.